src/LessMsi.Cli/Program.cs

```diff
@@ -95,25 +101,81 @@ public static int Main(string[] args)
+        /// </summary>
+        /// <param name="msiFileName">The path of the specified MSI file.</param>
+        /// <param name="outDirName">The directory to extract to. If empty it will use the current directory.</param>
+        /// <param name="filesToExtract">The files to be extracted from the msi. If empty all files will be extracted.</param>
+        /// /// <param name="extractionMode">Enum value for files extraction without folder structure</param>
+        public static void DoExtraction(string msiFileName, string outDirName, List<string> filesToExtract, eExtractionMode extractionMode)
+        {
+            if (string.IsNullOrEmpty(outDirName))
+                outDirName = Path.GetFileNameWithoutExtension(msiFileName);
+
+            EnsureFileRooted(ref msiFileName);
+            EnsureFileRooted(ref outDirName, extractionMode);
 
             var msiFile = new LessIO.Path(msiFileName);
 
-			Console.WriteLine("Extracting \'" + msiFile + "\' to \'" + outDirName + "\'.");
+            Console.WriteLine("Extracting \'" + msiFile + "\' to \'" + outDirName + "\'.");
 
-			Wixtracts.ExtractFiles(msiFile, outDirName, filesToExtract.ToArray(), PrintProgress);
-		}
+            Wixtracts.ExtractFiles(msiFile, outDirName, filesToExtract.ToArray(), PrintProgress);
+
+            if (isExtractionModeFlat(extractionMode))
+            {
+                Dictionary<string, int> fileNameCountingDict = new Dictionary<string, int>();
+
+                string originalOutDirName = outDirName.Remove(outDirName.Length - c_TempFolderSuffix.Length) + "\\";
+                Directory.CreateDirectory(originalOutDirName);
+
+                copyFilesInFlatWay(outDirName, originalOutDirName, extractionMode, fileNameCountingDict);
+
+                Directory.Delete(outDirName, true);
+            }
+        }
+
+        private static bool isExtractionModeFlat(eExtractionMode extractionMode)
+        {
+            bool flatExtractionModeFlag = extractionMode == eExtractionMode.OverwriteFlatExtraction;
+            flatExtractionModeFlag |= extractionMode == eExtractionMode.RenameFlatExtraction;
+
+            return flatExtractionModeFlag;
+        }
+
+        private static void copyFilesInFlatWay(string sourceDir, string targetDir, eExtractionMode extractionMode, Dictionary<string, int> fileNameCountingDict)
+        {
+            var allFiles = Directory.GetFiles(sourceDir);
+
+            foreach (var filePath in allFiles)
+            {
+                string fileSuffix = string.Empty;
+                string fileName = Path.GetFileName(filePath);
+
+                if (extractionMode == eExtractionMode.RenameFlatExtraction)
+                {
+                    if (fileNameCountingDict.ContainsKey(fileName))
+                    {
+                        fileSuffix = $"_{fileNameCountingDict[fileName]}";
+                        fileNameCountingDict[fileName]++;

```

Comment:
Double check this line works the way you expect. Is the int value returned by the dict's get method returned by reference (I forget how .net works in this case)?

Reply:
Before sending this pull request, I did some manual testing with NUnit msi file.
I didn’t encounter any problems with renaming identical files with their appearance count.